packages/typescript-estree/src/convert.ts

```diff
@@ -787,6 +788,14 @@ export class Converter {
       }
 
       case SyntaxKind.Identifier: {
+        if (isThisInTypeQuery(node)) {
```

Comment:
Changing the AST feels hacky and is gonna break some plugins. The `no-undef` rule uses `globalScope.through` to check undefined variables. Creating a variable called `this` for functions should prevent eslint from reporting this.


https://github.com/eslint/eslint/blob/main/lib/rules/no-undef.js#L59L75

Reply:
this looks like an issue in typescript, we should either get `ts.ThisType` or `ts.ThisExpression`

[playground](https://typescript-eslint.io/play/#ts=4.5.4&sourceType=module&showAST=ts&code=C4TwDgpgBAHlC8VSQPYDMkAsCWBnAsAFBEDGKAdrsFAIYBcWeCUATEA&rules=N4XyA&tsConfig=N4XyA)

Reply:
older versions of typescript reported `typeof this` as error `Cannot find name 'this'.`

Reply:
as for parsing, this can be done in way more efficient way, this issue is present only in `typeof` syntax, than we could validate this in `TypeQuery` instead to not have to iterate over all parents